### ISSUE TITLE
Android Studio: Problem with cmake/host and Protobuf of selected version (3.0.0-p1)

### ISSUE THREAD

**Issue by bezigon** [ruslo/hunter#1681]
* I've read [Brief overview section](https://docs.hunter.sh/en/latest/overview.html) and do understand basic concepts. **Yes**
* I've read [F.A.Q. section](https://docs.hunter.sh/en/latest/faq.html) and there is no solution to my problem there. **Yes**
* I've read [Code of Conduct](https://github.com/ruslo/hunter/blob/master/.github/CODE_OF_CONDUCT.md), I promise to be polite and will do my best at being constructive. **Yes**
* I've read [Reporting bugs section](https://docs.hunter.sh/en/latest/contributing.html#reporting-bugs) carefully. **Yes**
* I've checked that all the `hunter_add_package`/`find_package` API used by me in example is the same as in [documentation](https://docs.hunter.sh/en/latest/packages.html). **Yes**
* I'm using latest Hunter URL/SHA1. **Yes**

<!--- * I've created SSCCE reproducing the issue:
```cmake
# CMakeLists.txt

cmake_minimum_required(VERSION 3.0)

include("cmake/HunterGate.cmake")
HunterGate(
    URL "..."
    SHA1 "..."
)

project(foo)

hunter_add_package(boo)
find_package(boo CONFIG REQUIRED)
``` -->

<!--- If it's not just one file and you have to add more CMakeLists.txt or some C++ sources -->
<!--- then it's better to create separate GitHub repo with example -->
* I've created SSCCE GitHub repository to reproduce the issue: https://github.com/bezigon/android-studio-with-hunter

<!--- Link to CI jobs to reproduce an issue. Optional, but nice to have one. -->
<!--- * AppVeyor build reproducing issue: https://ci.appveyor.com/project/<username>/hunter/build/<build-number>
* Travis build reproducing issue: https://travis-ci.org/<username>/hunter/builds/<build-number> -->

Here is the log until first error reported by Hunter, option `HUNTER_STATUS_DEBUG` is `ON`:
* [Terminal Saved Output.log](https://github.com/ruslo/hunter/files/2694007/Terminal.Saved.Output.log)


<!--- Check this document: https://github.com/ruslo/hunter/wiki/error.external.build.failed -->
<!--- If you got `external.build.failed` -->
I've checked that the first error in logs **IS NOT** `external.build.failed`. **Yes**

<!--- Info about environment -->
* I'm building on OSX in Android Studio/Gradle.
* I'm using brewed CMake release (`brew install cmake`)
* Android Studio version: `3.2.1`
* Gradle version: `5.0`
* CMake version: `3.13.2`

<!--- What toolchain you're using if any -->
* I'm not using toolchain
<!--- * I'm using unmodified toolchain <toolchain-name> from [Polly](https://github.com/ruslo/polly)
* I've patched toolchain <toolchain-name> from [Polly](https://github.com/ruslo/polly). Here is my patch: https://github.com/<username>/polly
* I'm using my toolchain: https://github.com/<username>/<repo-with-toolchain> -->

I'm using next command line on generate step:
```
gradle assembleDebug --info
```

---
<!--- END -->

**Comment by ruslo** [comment]
> Problem with cmake/host and Protobuf

What is the problem exactly? I don't see any Hunter/CMake/build errors

**Comment by bezigon** [comment]
In fact, the problem is that the process does not go beyond after specific line `Scanning dependencies of target Protobuf-Release` in the near end of the log below.

```
[== "CMake Server" ==[

{"cookie":"","inReplyTo":"configure","message":"[hunter] \n************************* !!! WARNING !!! *************************\n\nYou're using experimental feature. This may not work well\nand can be removed anytime. It's a workaround for existing\nfeature request:\n\nhttps://github.com/ruslo/hunter/issues/495\n\n*******************************************************************","type":"message"}

]== "CMake Server" ==]

[hunter]
************************* !!! WARNING !!! *************************

You're using experimental feature. This may not work well
and can be removed anytime. It's a workaround for existing
feature request:

https://github.com/ruslo/hunter/issues/495

*******************************************************************
[hunter]
************************* !!! WARNING !!! *************************

You're using experimental feature. This may not work well
and can be removed anytime. It's a workaround for existing
feature request:

https://github.com/ruslo/hunter/issues/495

*******************************************************************


[== "CMake Server" ==[

{"cookie":"","inReplyTo":"configure","message":"[hunter] Building host project: /Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/cmake/host","type":"message"}

]== "CMake Server" ==]

[hunter] Building host project: /Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/cmake/host
[hunter] Building host project: /Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/cmake/host


[== "CMake Server" ==[

{"cookie":"","inReplyTo":"configure","message":"[hunter *** DEBUG *** 2018-12-19T12:22:55] Using default host generator","type":"message"}

]== "CMake Server" ==]

[hunter *** DEBUG *** 2018-12-19T12:22:55] Using default host generator
[hunter *** DEBUG *** 2018-12-19T12:22:55] Using default host generator
-- [hunter] Calculating Toolchain-SHA1

-- [hunter] Calculating Toolchain-SHA1
-- [hunter] Calculating Config-SHA1

-- [hunter] Calculating Config-SHA1
-- [hunter] HUNTER_ROOT: /Users/bezigon/.hunter

-- [hunter] HUNTER_ROOT: /Users/bezigon/.hunter
-- [hunter] [ Hunter-ID: c7b6099 | Toolchain-ID: 814ec5f | Config-ID: c9a4aa4 ]

-- [hunter] [ Hunter-ID: c7b6099 | Toolchain-ID: 814ec5f | Config-ID: c9a4aa4 ]
-- [hunter] PROTOBUF_ROOT: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Install (ver.: 3.0.0-p1)

-- [hunter] PROTOBUF_ROOT: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Install (ver.: 3.0.0-p1)
-- [hunter] Building Protobuf

-- [hunter] Building Protobuf
loading initial cache file /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/cache.cmake

loading initial cache file /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/cache.cmake
loading initial cache file /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Build/Protobuf/args.cmake

loading initial cache file /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Build/Protobuf/args.cmake
-- The C compiler identification is AppleClang 10.0.0.10001145

-- The C compiler identification is AppleClang 10.0.0.10001145
-- The CXX compiler identification is AppleClang 10.0.0.10001145

-- The CXX compiler identification is AppleClang 10.0.0.10001145
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc

-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works

-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compile features

-- Detecting C compile features
-- Detecting C compile features - done

-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++

-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works

-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compile features

-- Detecting CXX compile features
-- Detecting CXX compile features - done

-- Detecting CXX compile features - done
-- Configuring done

-- Configuring done
-- Generating done

-- Generating done
-- Build files have been written to: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Build/Protobuf/Build

-- Build files have been written to: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Build/Protobuf/Build
Scanning dependencies of target Protobuf-Release

Scanning dependencies of target Protobuf-Release

> Task :app:generateJsonModelDebug FAILED
:app:generateJsonModelDebug (Thread[Execution worker for ':',5,main]) completed. Took 23.305 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:generateJsonModelDebug'.
> Conversion = ']'

* Try:
Run with --stacktrace option to get the stack trace. Run with --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.0/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 25s
16 actionable tasks: 1 executed, 15 up-to-date
bezigon@bezigons-iMac ~/D/G/a/android-studio>
```


Of course, I deleted the `~/.hunter/_Base` directory for the purity of the experiment.

**Comment by bezigon** [comment]
In the project `protobuf-toolchains-mix` the assembly on pure CMake/OS X of selected version of Protobuf (3.0.0-p1) passes without problems.

**Comment by ruslo** [comment]
> In fact, the problem is that the process does not go beyond after specific line Scanning dependencies of target Protobuf-Release in the near end of the log below.

It doesn't mean that problem is in Hunter, it may be a Gradle + CMake issue.

Try to use workaround from documentation:
* https://docs.hunter.sh/en/latest/faq/android-studio.html#example
```
...
In this case you can add return() command right after first hunter_add_package call
...
```

Another example: https://github.com/elucideye/drishti#android-studio-workarounds

**Comment by bezigon** [comment]
Already did that feature using `return()`, but no success.

```bezigon@bezigons-iMac ~/D/G/a/android-studio> cmake --build app/.externalNativeBuild/cmake/debug/arm64-v8a
[0/1] /usr/local/Cellar/cmake/3.13.2/bin/cmake -S/Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon -B/Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/android-studio/app/.externalNativeBuild/cmake/debug/arm64-v8a
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] HUNTER_ROOT set using HOME environment variable
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] HUNTER_ROOT: /Users/bezigon/.hunter
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] Settings (initialize):
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   HunterGate done (NO)
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   Cache init (YES)
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] Single-configuration generator
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] Using default cache server
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] List of cache servers:
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   * https://github.com/ingenue/hunter-cache
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] Settings (finalize):
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   HunterGate done (NO)
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   Cache init (YES)
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] Variables from HunterGate:
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   HUNTER_GATE_ROOT: /Users/bezigon/.hunter
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   HUNTER_GATE_VERSION: 0.23.76
-- [hunter *** DEBUG *** 2018-12-19T13:02:12]   HUNTER_GATE_SHA1: c7b60993e841850e2c449afd454f5d5aa4ec04e4
-- [hunter] Calculating Toolchain-SHA1
-- [hunter *** DEBUG *** 2018-12-19T13:02:12] [/Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/android-studio/app/.externalNativeBuild/cmake/debug/arm64-v8a/_3rdParty/Hunter/toolchain] > "/usr/local/Cellar/cmake/3.13.2/bin/cmake" "-DTOOLCHAIN_INFO_FILE=/Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/android-studio/app/.externalNativeBuild/cmake/debug/arm64-v8a/_3rdParty/Hunter/toolchain/toolchain.info.NOLF" "-DCMAKE_TOOLCHAIN_FILE=/Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/android-studio/app/.externalNativeBuild/cmake/debug/arm64-v8a/generated/toolchain.cmake" "-DHUNTER_SELF=/Users/bezigon/.hunter/_Base/Download/Hunter/0.23.76/c7b6099/Unpacked" "-GNinja" "-H/Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/android-studio/app/.externalNativeBuild/cmake/debug/arm64-v8a/_3rdParty/Hunter/toolchain" "-B/Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/android-studio/app/.externalNativeBuild/cmake/debug/arm64-v8a/_3rdParty/Hunter/toolchain/_builds" "-DCMAKE_MAKE_PROGRAM=/usr/local/bin/ninja" "-DCMAKE_RELEASE_POSTFIX=" "-DCMAKE_DEBUG_POSTFIX=d"
-- Android: Targeting API '24' with architecture 'arm64', ABI 'arm64-v8a', and processor 'aarch64'
-- Android: Selected Clang toolchain 'aarch64-linux-android-clang' with GCC toolchain 'aarch64-linux-android-4.9'
-- Check for working C compiler: /Users/bezigon/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang
-- Check for working C compiler: /Users/bezigon/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Users/bezigon/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++
-- Check for working CXX compiler: /Users/bezigon/Library/Android/sdk/ndk-bundle/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/android-studio/app/.externalNativeBuild/cmake/debug/arm64-v8a/_3rdParty/Hunter/toolchain/_builds
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] Already exists: /Users/bezigon/.hunter/_Base/c7b6099/4e02b95/toolchain.info
-- [hunter] Calculating Config-SHA1
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] Cache data is up-to-date
-- [hunter] HUNTER_ROOT: /Users/bezigon/.hunter
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] HUNTER_TOOLCHAIN_ID_PATH: /Users/bezigon/.hunter/_Base/c7b6099/4e02b95
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] HUNTER_CONFIGURATION_TYPES: Release;Debug
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] HUNTER_BUILD_SHARED_LIBS:
-- [hunter] [ Hunter-ID: c7b6099 | Toolchain-ID: 4e02b95 | Config-ID: 733cc58 ]
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] load: /Users/bezigon/.hunter/_Base/Download/Hunter/0.23.76/c7b6099/Unpacked/cmake/projects/md5/hunter.cmake
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] md5 versions available: [1.6]
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] Package 'md5' CONFIGURATION_TYPES: Release;Debug
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] Package 'md5' is cacheable: YES
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] Install to: /Users/bezigon/.hunter/_Base/c7b6099/4e02b95/733cc58/Install
-- [hunter] MD5_ROOT: /Users/bezigon/.hunter/_Base/c7b6099/4e02b95/733cc58/Install (ver.: 1.6)
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] Package already installed: md5
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] load: /Users/bezigon/.hunter/_Base/Download/Hunter/0.23.76/c7b6099/Unpacked/cmake/projects/md5/hunter.cmake ... end
-- [hunter]
************************* !!! WARNING !!! *************************

You're using experimental feature. This may not work well
and can be removed anytime. It's a workaround for existing
feature request:

https://github.com/ruslo/hunter/issues/495

*******************************************************************
-- [hunter] Building host project: /Users/bezigon/Documents/GitHub/android-studio-with-hunter-bezigon/cmake/host
-- [hunter *** DEBUG *** 2018-12-19T13:02:18] Using default host generator
-- [hunter] Calculating Toolchain-SHA1
-- [hunter] Calculating Config-SHA1
-- [hunter] HUNTER_ROOT: /Users/bezigon/.hunter
-- [hunter] [ Hunter-ID: c7b6099 | Toolchain-ID: 814ec5f | Config-ID: c9a4aa4 ]
-- [hunter] PROTOBUF_ROOT: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Install (ver.: 3.0.0-p1)
^Cninja: error: rebuilding 'build.ninja': interrupted by user```

**Comment by ruslo** [comment]
```
-- [hunter] Calculating Toolchain-SHA1
-- [hunter] Calculating Config-SHA1
-- [hunter] HUNTER_ROOT: /Users/bezigon/.hunter
-- [hunter] [ Hunter-ID: c7b6099 | Toolchain-ID: 814ec5f | Config-ID: c9a4aa4 ]
-- [hunter] PROTOBUF_ROOT: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Install (ver.: 3.0.0-p1)
```

`HUNTER_STATUS_DEBUG=ON` is missing on host level, can you please add `option(HUNTER_STATUS_DEBUG "..." ON)` to host `CMakeLists.txt`?

**Comment by ruslo** [comment]
tl;dr I suspect that something is going wrong

> Added exit() feature strictly after hunter_add_package(Protobuf)

Since `Protobuf` is problematic part I do recommend you (at least for debugging for now) to add something plain simple, e.g. `md5`.

> Performed gradle assembleDebug -Parch=arm64-v8a --info target
> Waited several time on
> and forcefully exited

Part with "early exit" should be completed succesfully using Gradle, if it's not competed then something went wrong and workaround is not working.

```
-- [hunter] Calculating Toolchain-SHA1
-- [hunter] Calculating Config-SHA1
-- [hunter] HUNTER_ROOT: /Users/bezigon/.hunter
-- [hunter] [ Hunter-ID: c7b6099 | Toolchain-ID: 814ec5f | Config-ID: c9a4aa4 ]
-- [hunter] PROTOBUF_ROOT: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Install (ver.: 3.0.0-p1)
```

I don't see `HUNTER_STATUS_DEBUG=ON` was applied here...

**Comment by bezigon** [comment]
I already used `md5` in parallel with `Protobuf`, please look at https://github.com/bezigon/android-studio-with-hunter

The part with `md5` passes without problems.

It's interesting that the `protobuf` of the latest version (`3.5.2`) runs without problems, though I had to wait a while.

I will try to use the latest version, although there are times when a specific protobuf version is required... Thanks anyway for your support.

**Comment by ruslo** [comment]
> I already used md5 in parallel with Protobuf, please look at https://github.com/bezigon/android-studio-with-hunter

That's not what I see:

>    Waited several time on

> -- [hunter] PROTOBUF_ROOT: /Users/bezigon/.hunter/_Base/c7b6099/814ec5f/c9a4aa4/Install (ver.: 3.0.0-p1)
> <=<======-------> 50% EXECUTING [1m 28s]
> > > :app:generateJsonModelDebug
> ^C?

> and forcefully exited

Meaning you have `hunter_add_package(protobuf)` call, I do expect that we exit **before**. Should be:

```cmake
hunter_add_package(md5)
return() # workaround

hunter_add_package(protobuf) # not used in workaround
```

So I'm still not sure that you're using workaround correctly.

> It's interesting that the protobuf of the latest version (3.5.2) runs without problems, though I had to wait a while.

Sources are here: https://github.com/hunter-packages/protobuf feel free to check

**Comment by bezigon** [comment]
I understood you. Will retry with workaround on first occurence of `hunter_add_package(...)`.
